%\VignetteIndexEntry{mmnet}
%\VignetteKeywords{mmnet}
%\VignettePackage{mmnet}

\documentclass[a4paper]{article}
\usepackage{times}
\usepackage{natbib}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{Sweave}
\usepackage[utf8]{inputenc}
% \usepackage[pdftex]{graphicx}
\usepackage{url} 

\title{mmnet: Metagenomic analysis of microbiome metabolic network}
\author{Yang Cao, Fei Li, Xiaochen Bo}

\begin{document}
\bibliographystyle{plainnat}

\SweaveOpts{highlight=TRUE, tidy=TRUE, keep.space=TRUE, keep.blank.space=FALSE, keep.comment=TRUE}
\maketitle
\tableofcontents

\section{Introduction}

This manual is a brief introduction to structure, functions and uasge of \emph{mmnet} package. It demonstrates a metagenomic pipline for systems biology, from raw metagenomic data annotation to network analysis. The \emph{mmnet} package provides an implementation of metagenome system analysis of metagenomic data in R, utilizes a metagenomic systems biology computational framework mentioned in \citep{greenblum2012metagenomic}.

This package addresses the microbiome metabolic network construction and analysis. Below we lists the main features of \emph{mmnet}: 

  \begin{itemize}
    \item Building a local KEGG orthology dataset 
    \item Constructing the reference metabolic networkf
    \item Annotation of metagenomic sequence reads with Metagenomics RAST sever 
    \item Estimating the abundance of enzymatic genes
    \item Constructing State Specific Network (SSN)
    \item SSN topological ananlysis 
    \item Abundance comparison of enzymatic genes
  \end{itemize}

\subsection{Installation}

\emph{mmnet} requires that \emph{KEGGREST}, \emph{igraph}, \emph{Biobase}, \emph{XML}, \emph{RCurl}, \emph{RJSONIO}, \emph{stringr}, \emph{ggplot2}, \emph{biom} are installed. These should be installed automatically when you install \emph{mmnet}. 

Note:  the latest stable release version of the biom package just 0.3.12, \emph{mmnet} requires  \emph{biom} must be 0.3.13 or later, uses should install the latest development version of the biom package, enter the following lines in an R session:

<<biom, eval=FALSE>>==
install.packages("devtools") # if not already installed
devtools::install_github("biom", "joey711")
@

For more functionally, please install the latest devel version \emph{mmnet}, use the devel version of bioconductor:
<<install-pkg, eval=FALSE>>==
if (!match("BiocInstaller", installed.packages)){
  source("http://bioconductor.org/biocLite.R")
  biocLite("BiocInstaller")
}
library(BiocInstaller)
useDevel()
biocLite("mmnet")
@

User can load the \emph{mmnet} to the current workspace by typing or pasting the following codes in R command line:

<<load-pkg,results=hide,eval=TRUE>>==
library(mmnet)
@

subsection{Further help} 
To view the \emph{mmnet} description and a summary of all the functions within \emph{mmnet} type the following:

<<help-pkg,results=hide,eval=FALSE>>==
library(help = mmnet)
@

Thanks for using this package. Please contact us if there are any bugs or suggustiones.

\section{A use case: from raw sequence data to metabolic network analysis}

We will demonstrate some of the functions in \emph{mmnet}. A public dataset which have been annotated in MG-RAST was taken for demenstration purposes since the time metagenomic sequence taken on MG-RAST will range between a few hours and weeks. There is only a  simple code example for microbiome functional annotation.

The test data is consisting of several microbiomes from twin pairs and their mothers and available in MG-RAST. 

\subsection{Local KEGG orthology dataset and reference metabolic network constructing}

This package takes KEGG database to annotate enzymatic genes with metabolic reactions which is the basis representation for all proteins and functional RNAs corresponding to KEGG pathway nodes, BRITE hiearchy nodes, and KEGG module nodes to annotation the microbiome.  The KEGG metabolic pathway is the best organized part of KEGG PATHWAY database, and also is a network of KO-KO  relations \citep{kanehisa2000kegg}. It is composed of KO, substrate and product of KO, and have been applied widely to create genome-scale metabolic networks of various microbial species \citep{oberhardt2009applications}. \emph{This small reference data was obtained with the KEGG free REST API.}

An initial reference data named \textit{RefDbcache.rda} was saved in "data" subdirectory of \emph{mment}, that contains KOs and anntated with a metabolic reaction. Morever, an KO based reference metabolic network was also saved in \textit{RefDbcache.rda}, where nodes represent enzymes (KOs), and the directed edge from enzyme A to enzyme B indicates that a product metabolite of a reaction catalyzed by enzyme A is a substrate metabolite of a reaction catalyzed by enzyme B. Howerver, KEGG database updated frequently, reference data can be updated by function \textit{updateMetabolicNetwork} manually and saved in the directory user specified. 

<<initial-data,echo=TRUE>>=
## the initial saved reference data
   data(RefDbcache)
##
   names(RefDbcache)
## enzymatic genes (KOs) information, metabolites of annotated reation is 
## saved as substrate and product 
   head(RefDbcache$ko)
## the reference network
   RefDbcache$network
@

<<update-kegg,echo=TRUE>>==
## the updated data can be saved in your workongspace defaultly
# updateKEGGPathway(path = Sys.getenv("HOME"))
@

Each KO is assigned a set of substrate and product metabolites according to the reactions catalyzed by the KO as annotated in KEGG. A search is performed to identify all KO pairs in which a product metabolite of one KO is a substrate metabolite of the other. Directed network edges are then drawn between each identified KO pair.

<<referenceNet,echo=TRUE>>==
## reference network construction saved as igraph object
# require(igraph)
# refNet <- constuctMetabolicNetkwrok(RefDbcache) 
@

\subsection{Annotation of metagenomic sequence reads}

Functional annotation of microbiome is a crucial step for subsequent analysis. \emph{mmnet} provides functions from user login to sequence annotation linking \emph{R} with MG-RAST (Metagenomics RAST) server \citep{glass2010using}. MG-RAST is an automated analysis platform for metagenomes providing a new paradigm for the annotation and analysis of metagenomes based on sequence data, and is also stable, extensible, and freely available to all researchers. Before annatated microbiome, your should register an account in MG-RAST. Fore more details see \url{http://metagenomics.anl.gov}. 

Several functions were used for sequence annotation on MGRAST in \emph{mmnet}, \textit{loginMgrast},\textit{uploadMgrast}, \textit{submitMgrastJob}, \textit{checkMgrastProject}. Function \textit{runMgrastJob} is a wrapper for the functions for annotation in the \emph{mmnet} package, which achieves processing from sequence uploading to annotation profile dowloading, and will return the annotatation profile after annotation complete on MGRAST. 

Note: time sequence annotation took will range from days to weeks.

<<runMgrastJob,echo=TRUE>>==
## provide the MG-RAST account,rawsequence, and the project name you will creat, 
## annotation of your sequence iss automatically processing on MG-RAST 
## here will return a MG-RAST job id
# runMgrastAnnotation(seqfiles,user,userpwd,new_project)
@

Moreover, function \textit{checkMgrastProject} is used to show all private projects completely calculated on MG-RAST anytime as well as jobs still in calculation. 

To check your MG-RAST job status, we have to call \textit{loginMgrast} to login MGRAST first. See \texttt{'?loginMgrast'} for details. It returns the login information (contains webkey, websession etc) as an authenticate of subsequent operation on MG-RAST.

<<checkProject-mgrast, echo=TRUE>>==
## loginMgrast help user login MG-RAST if not login
# login.info <- loginMgrast(user="mmnet",userpwd="mmnet")
# names(login.info)

##
#project.status <- checkMgrastProject(login.info)
@
 
\emph{mmnet} also provides a function \textit{getMgrastAnnotation} to load the functional annotation profile (both priveta and public) separately for metabolic analysis.

<<getannotation,echo=TRUE,results=hide>>==
## public annotation data, 
# publicAnnotation <- getMgrastAnnotation("mgm4440824.3")

## privete annotation data needs your webkey on MG-RAST
# webkey <- login.info["webkey"]
# privateAnnotation <- getMgrastAnnotation("mgm4440824.3"), 
# webkey)
@

\subsection{Estimating the abundance of enzymatic genes}

As mentioned above, a public dataset \citep{turnbaugh2008core} was taken for example data, four samples of mpg10 (metagenome project 10) in MGRAST. Apparently, user could obtain the annatated information with function \textit{getMgrastAnnotation}. For saving claculating time, we have saved it as \textit{mpg10Anno.rda} in \emph{mmnet}. It can be loaded by typing or pasting the following:

<<testData,echo=TRUE>>==
## the test data from the dataset mpg10 in MG-RAST, four microbiome, 
# two obese and two lean
  data(mgp10Anno)
  summary(mgp10Anno)
@

To estimate the abundance of enzymatic genes, \textit{estimateAbundance} function should be called for each sample. A widely used in metagenomic analys format - Biological Observation Matrix (BIOM) \citep{mcdonald2012biological} format will be return by this function. The BIOM file format (canonically pronounced biome) is designed to be a general-use format for representing biological sample by observation contingency tables. More details on BIOM file can be found in \url{http://biom-format.org/}  R package \emph{biom}

<<estimate,echo=TRUE>>==
 KOAbund <- estimateAbundance(mgp10Anno)
 show(KOAbund)
@

For comparative analysis among samples with different state (e.g. obese or lean), correlation coefficient was used to measure relative KO abundance across the samples.

<<correlation,echo=TRUE>>==
## pearson correlation
 data <- biom_data(KOAbund)
 cor(as.matrix(data),method = "pearson")
@
 Appanrently, KO abundance across all samples was highly concordant (R>0.77)

 \subsection{Constructing State Specific Network}

% The \emph{mmnet} provides function \emph{mapAnnotoKEGGPathway} for mapping the KOs  on KEGG global metabolic pathway. This function returns an url of colored map, users may automated open the url by optional "browse" or manually open it.

% <<showKEGGPathway,echo=TRUE>>==
% colored.map <- showKEGGPathway(names(KOAbund[[1]]), map.color="blue")
% @

% The colored KEGG global metabolic map with example metagenomic data.

% \begin{figure}[h]
% \centering\includegraphics{2ec01100.png}
% \caption{A example colored KEGG global map}\label{fig:01}
% \end{figure} 

% The colored KEGG global metabolic map with example metagenomic data.

Different microbiomes with various states (obese or lean) have different KO abundances. We provides function \textit{stripSSN} to strip the reference netwok to obtain the state specific metabolic network for each microbiome. It could take the output of function \emph{estimateAbundance} as input. Moreover, while samples has been annotated with other tools (e.g. blast to KEGG locally), users could create their own BIOM file represents enzymatic genes abundance. The KO abundance was add to the network as a nodes attributes automatically. 

<<strip-network,echo=TRUE>>==
  ssn <- stripSSN(KOAbund)
  g <- ssn[[3]]
  summary(g)
  abund <- get.vertex.attribute(g,"abundance",index=V(g))
  summary(abund)
@

\subsection{Topogolical analysis of SSN and abundance comparison of enzymatic genes}

 Function \textit{analyzeSSN} links the difference KO abundance with the topological features of state specific metabolic network to find the topological differences among SSNs. Five most presented vertex topological feartures is provided by this function, added to the newtwork as nodes attributes, it also can plot the SSN. See \texttt{'?analyzeSSN'} for details. 

\begin{figure}[!htbp]
  \begin{center}
<<analyzeSSN, echo=TRUE, fig=TRUE>>==
 topo.feature <- analyzeSSN(g)

 ## network with topological features as attributes
 topo.feature
@
\caption{Scatterplot between abundance and topological features}\label{fig:01}
  \end{center}
\end{figure}

% \begin{figure}[!htbp]
%   \begin{center}
% <<echo=TRUE, fig=TRUE>>==
% topo.feature$scatterplot
% @
% \caption{A metabolic network for single sample, node size is positive proportional log(abundance)}\label{fig:02}
%   \end{center}
% \end{figure}

To compare the abundance of enzymatic genes (KOs) across various samples, we take three strategies odds ratio, difference rank of each KO or JSD (Jensen-Shannon Divergence) to identify the abnormal KO (enrich or deplete). It will construct a compararive network, which nodes of these abnormal KO with the specific color by calling function \emph{compareKOState}, it also can plot the compared network.

\begin{figure}[!htbp]
  \begin{center}
<<compareState, echo = TRUE, fig = TRUE>>==
 comparedNet <- compareSSN(KOAbund, method="OR", cutoff = c(0.5, 2))
 summary(comparedNet)
@
\caption{A metabolic network for multiple sample, KOs that are associated with specific state appear as colored nodes (red=enriched, green=depleted)}\label{fig:02}
  \end{center}
\end{figure}

Subsequently, users can explore the correlations between the topological features and the differential abundance with function \textit{analyzeSSN}:

<<analyzeSSN2, echo=TRUE, fig=FALSE>>==
 abund.compare <- analyzeSSN(g)

 ## network with topological features as attributes
 topo.feature
@

\subsection{Visualization using RCytoscape}

Both function \textit{analyzeSSN} and \textit{compareKOState} called function \textit{showSSN} to plot the metabolic network using the functions available in the \emph{igraph} package above, means that \textit{showSSN} can be used to personalized network display by specificing appropriate parameters.

\begin{figure}[!htbp]
  \begin{center}
<<showSSN,echo=TRUE,fig=TRUE>>==
## the reference network
# showSSN(RefDbcache$network,mode="ref")
#the state specific metabolic network
showSSN(g, mode="strip", vertex.label = NA, 
        edge.width = 0.3, edge.arrow.size = 0.1, edge.arrow.width = 0.1, 
        layout = layout.fruchterman.reingold, vertex.size = 3)

@
\caption{A metabolic network for single sample, node size is positive proportional log(abundance)}\label{fig:03}
\end{center}
\end{figure}

It is also possible to visualize the network using the functions available in the \emph{RCytoscape} package \citep{shannon2013rcytoscape}, showing metabolic network in Cytoscape. See\emph{RCytoscape} package for details on how to use transfer the network and attributes from R to Cytoscape.

Here is a simple example to show the reference metabolic network in Cytoscape with \emph{RCytoscape} package.

<<RCytoscape, eval = FALSE, echo = TRUE>>==
 if (require(RCytoscape)){
   refnet <- RefDbcache$network
   refnet <- igraph.to.graphNEL(refnet)
   ## initialize the edge attibute
   refnet <- initEdgeAttribute(refnet, attribute.name = "weight", 
            attribute.type = "char", default.value = "0")
   ## create a network window in Cytoscape
   cw <- new.CytoscapeWindow ('refnet', graph = refnet, overwriteWindow = TRUE)
   ## transmits the CytoscapeWindowClass's graph data, from R to Cytoscape, nodes, 
   ## edges, node and edge attributes
   displayGraph (cw)
 }
@

\begin{figure}[h]
\centering\includegraphics{Refnet.png}
\caption{Show reference metabolic netowork in Cytoscape}\label{fig:04}
\end{figure} 


% \begin{figure}[!htbp]
%   \begin{center}
% <<showSSN2,echo=TRUE, fig=TRUE>>=
% ## the compared network
% showSSN(comparedNet,mode="compared",method="OR")
% @
% \caption{A metabolic network for multiple sample, KOs that are associated with specific state appear as colored nodes (red=enriched, green=depleted).}\label{fig:03}
% \end{center}
% \end{figure}

\section{Futher Directions}

 Only metagenomic data have been annotated on MGRAST server  could this current version \emph{mmnet} analyze. We plan to achieve the metagenomic data annotation and integrate function for other annotation analysis (e.g. blast).

Another limitation is the  topological analysis of the network is also not enough precise because of the samll sample. Finally, the network constructing is time consuming for large-scale metagenomic data. This problems may be fixed to improve this package in the future, and we believe it will increase the practically of \emph{mmnet}.

\section*{Session Information}

The version number of R and packages loaded for generating the vignette were:

<<echo=FALSE>>=
sessionInfo()
@

\section*{Cleanup}


This is a cleanup step for the vignette on Windows; typically not
needed for users.

<<closeConnetions,results=hide>>=
allCon <- showConnections()
socketCon <- as.integer(rownames(allCon)[allCon[, "class"] == "sockconn"])
sapply(socketCon, function(ii) close.connection(getConnection(ii)) )
@

\bibliography{mmnet}

\end{document}